### ISSUE TITLE
Proposal - A better way of managing errors through the GoodJob UI

### ISSUE THREAD

**Issue by lucasfcunha** [bensheldon/good_job#1464]
Hi,

We recently started migrating our background jobs from Resque to GoodJob. We are very happy with the change overall, the only thing lacking a bit is the UI for managing errors, it's hard to get a good overall picture of what failed, which error happened, and when.

I would like to know if @bensheldon would be willing to accept a contribution to the GoodJob UI to create a "cleaner" tab on the UI that is similar to what Resque had.

![SCR-20240809-kjhn](https://github.com/user-attachments/assets/1ae06ff6-aada-4c24-b4cb-3effa9cbb064)

The idea would be to create a new tab beside the "Performance" tab at the top called "Cleaner" or something similar that would house a UI that resembles the one from Resque above. A table indicating which queues had errors and separate when those happened with some intervals as well as the same view based on the type of error that happened (exception fired).

To achieve this we would have also to create a new column on the table `good_jobs` called `error_class` to filter the jobs that failed with a specific exception. This could become a specific filter on the discard page as well.

The links on the table would drive the user to the "discarded" tab with that specific timeframe/error/class filtered, so they could just retry the jobs or destroy it if necessary.

I believe this change would have a good impact on people who run a lot of different jobs in a production environment and need to have a better visualization of errors happening.

@bensheldon let me know what you think about this, if you think that this is positive and you would be willing to merge it, we would be happy to contribute on this.

**Comment by bensheldon** [comment]
I'm ok with this generally. I think it could simply be called "Errors" in the top navbar.

I think the design questions that come to mind are:

- Generally, how does one use this view? I think seeing a list of errors, and a relative frequency/pain/impact is helpful, but like what's the human workflow one uses this for? Just curious how you use this page with Resque (I'm guessing "Cleaner" implies a usage I'm not familar with). For context, one of my goals for GoodJob is to reduce the effort of using/maintaining a background job system, so I'm wondering (generally) if there are specific ops/anxiety-led question(s) that would lead to someone checking this view, that we could more directly respond to than providing a table of values to eyeball (not a blocking musing).
- What in the job lifecycle should this show? Should an error only show up here once the job has been discarded, or should this also be errors for each and every execution?
- Could the end-result be achieved instead by added an "error class" dropdown filter on the main jobs page instead, and that would utilize the chart to display similar data.

In terms of implementation, not too much sweat adding a new column and an index (probably `(error_class, job_class)` for grouping). Though I wonder if this could be done (albeit less efficiently) with the existing `error` column (Postgres can split/group the current `error` column on the `: ` separator, though via a table scan). Or grouping on the entire `error` column, though it would fall apart on high-cardinality error messages (or handle that in Ruby). It's easiest to ship and iterate without schema changes.

In summary:

- Sure 👍 
- Can you prototype it without making a schema change?
- I'm curious how you use this table in your regular ops/developer practices.

**Comment by lucasfcunha** [comment]
Answering some of your questions

1) We process ~500k+ daily jobs on our infrastructure. Even though we have retry logic for most of the jobs, sometimes things fail and go to the discarded pile. We use this screen on Resque today to understand the big picture of what's happening, if we need to retry a specific error/class it's easy to go to the table and click on the one you want to take some action for example. Today there is no easy way on the GoodJob UI to understand how many jobs were discarded in a specific queue, you need to manually filter on the discarded page to see the count.
2) They would only show up on that page when they were discarded, those are things that someone needs to manually take some action.
3) It could, but we would not still see the "big picture", because of the amount and different queues/jobs we have, it's time-consuming doing that one per one.

We can do that without an `error_class` column, but I believe it would not be very efficient.

Thanks for the quick response @bensheldon. We are at the moment in the middle of migrating to GoodJob, but once we have a little bit of a breather we will work on a prototype of this page.

**Comment by Earlopain** [comment]
I think showing both discards and retries makes sense. Our jobs seldom fail totally but there are some intermittent issues that pop up occasionally that are handled via retries.

Can be a separate tab or something. The implementation will be very similar, just pulling data from a different table. If you're not interested in that its all good, I can add it myself afterwards if I feel like doing it. Your initial proposal is already nice.

----

Re colums: if a new `error_class` gets added, `error_message` seems to make sense as well to eventually drop the mixed `error` column.
